packages/ui/lib/components/Form/AddressInput.tsx

```diff
@@ -0,0 +1,179 @@
+ * @returns Input component
+ *
+ */
+ export const AddressInput = forwardRef(
+  (props: Props, ref: ForwardedRef<HTMLInputElement>) => {
+    const {
+      size = 'medium',
+      value,
+      className,
+      description,
+      label,
+      withIcon = true,
+      isTruncated,
+      web3Service,
+      localForm,
+      name,
+      ...inputProps
+    } = props
+
+    if (!localForm) return null;
+    const { register } = localForm
+  
+    const [addressType, setAddressType] = useState('')
+    const [resolvedAddress, setResolvedAddress] = useState('')
+    const [resolvedName, setResolvedName] = useState('')
+    const [loadingResolvedAddress, setLoading] = useState(false)
+    const [error, setError] = useState<any>('')
+    const [success, setSuccess] = useState('')
+
+    const inputSizeStyle = SIZE_STYLES[size]
+    let inputStateStyles = ''
+
+    if (error) {
+      inputStateStyles = STATE_STYLES.error
+    } else if (success) {
+      inputStateStyles = STATE_STYLES.success
+    }
+
+    const inputClass = twMerge(
+      'block w-full box-border rounded-lg transition-all shadow-sm border border-gray-400 hover:border-gray-500 focus:ring-gray-500 focus:border-gray-500 focus:outline-none flex-1 disabled:bg-gray-100',
+      inputSizeStyle,
+      inputStateStyles,
+      withIcon ? 'pl-10' : undefined
+    )
+
+    const handleResolver = async (address: string) => {
+      if (address.length) {
+        setLoading(true)
+      }
+      const result = await web3Service.resolveName(address)
+      if (result && result.type === 'address') {
+        setLoading(false)
+        setAddressType(result.type)
+        setSuccess(`It's a valid eth address`)
+        setResolvedName(result.name)
+        return result.address;
+      } else if (result && result.type === 'name') {
+        setLoading(false)
+        setAddressType(result.type)
+        setSuccess(`It's a valid ens name`)
+        setResolvedAddress(result.address)
+        return result.address;
+      } else {
+        setLoading(false)
+        setAddressType(result.type)
+        setError(`It's not a valid ens name or address`)
+        setSuccess('')
+        setResolvedAddress('')
+        return result.error;
+      }
+    }
+
+    return (
+      <>
+        <FieldLayout
+          label={label}
+          size={size}
+          success={success}
+          error={error}
+          description={description}
+        >
+          <div className="flex flex-col">
+            <div className="relative">
+              {withIcon && (
+                <span className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
+                  <Icon size={size} icon={CustomizedIcon} />
+                </span>
+              )}
+              <input
+                {...inputProps}
+                id={label}
+                className={inputClass}
+                {...register(name, {
+                  setValueAs: (value: string) => handleResolver(value),
```

Comment:
Wouldn't a debounce be useful given user will likely be typing an ens and each hit will fire a request to resolve? Does this not cause an issue currently? 

Also setValueAs is run after all validations passes instead of before so one cannot add any kind of validation on the resolved value. Do you think you can use onChange which runs before any validation happens?

Reply:
I will try adding the debounce function but regarding the latter, the goal is to pass the `resolvedAddress` back to parent, and as far as I have tried, `setValueAs` is the only way I was able to do it. And to be honest I don't see any scenarios where anyone would want to run validation on the `resolvedAddress`. It's a very edge-case scenario which I think we can add later.

Reply:
Hah. I actually do this with current places where ENS input is used which is why I explicitly mentioned it if this component were to replace them.